hera_qm/firstcal_metrics.py

```diff
@@ -0,0 +1,533 @@
+            firstcal delay solutions
+
+        self.delay_offsets : ndarray, shape=(N_ant, N_times)
+            firstcal delay solution offsets from time average
+
+        self.frac_JD : ndarray, shape=(N_times,)
+            ndarray containing time-stamps of each integration
+            in fraction of current JD
+
+        self.ants : ndarray, shape=(N_ants,)
+            ndarray containing antenna numbers
+        """
+
+        # Instantiate UVCal and read calfits
+        self.UVC = UVCal()
+        self.UVC.read_calfits(calfits_file)
+
+        # get file prefix
+        self.file_stem = '.'.join(calfits_file.split('.')[:-1])
+
+        # Calculate median delay
+        self.delays = self.UVC.delay_array.squeeze() * 1e9
+        self.delay_avgs = np.median(self.delays, axis=1)
+        self.delay_offsets = (self.delays.T - self.delay_avgs).T
+
+        # get other relevant arrays
+        self.times = self.UVC.time_array
+        self.start_JD = np.floor(self.times).min()
+        self.frac_JD = self.times - self.start_JD
+        self.minutes = 24 * 60 * (self.frac_JD - self.frac_JD.min())
+        self.Nants = self.UVC.Nants_data
+        self.ants = self.UVC.ant_array
+
+    def run_metrics(self, std_cut=0.5, output=False):
+        """
+        Run all metrics and write results to file
+
+        filename : str, default=None
+            filename for output w/o filetype suffix
+            if None, uses calfile stem
+
+        std_cut : float, default=0.5
+            delay stand. dev cut for good / bad determination
+
+        output : bool, default=False
+            return with function
+
+        filetype : str, default='pkl'
+            output filetype
+            options = ['pkl', 'json']
+
+        write : bool, default=True
+            write output file or not
+
+        Output:
+        -------
+        Create a self.result dictionary
+        if output == True:
+            return self.result
+
+        """
+        # Calculate std and zscores
+        self.ant_std, self.time_std, self.agg_std, self.z_scores = self.delay_std(return_dict=True)
+
+        # Given delay std cut find "bad" ant
+        # determine if full sol is bad
+        if self.agg_std > std_cut:
+            self.full_sol = 'bad'
+        else:
+            self.full_sol = 'good'
+
+        self.bad_ants = []
+        for ant in self.ant_std:
+            if self.ant_std[ant] > std_cut:
+                self.bad_ants.append(ant)
+
+        # put into dictionary
+        result = OrderedDict()
+        result['full_sol'] = self.full_sol
+        result['bad_ant'] = self.bad_ants
+        result['z_scores'] = self.z_scores
+        result['ant_std'] = self.ant_std
+        result['time_std'] = self.time_std
+        result['agg_std'] = self.agg_std
+        result['times'] = self.times
+        self.result = result
+
+        if output == True:
+            return self.result
+       
+    def write_metrics(self, filename=None, filetype='pkl'):
```

Comment:
Also, to load in a JSON and turn it into a dictionary it seems like we need to use the ```eval``` function which is something I try to never do. Pickle just seems a whole lot easier to deal with if we are going to be writing data structures, like dictionaries and ndarrays.